File: 공부/구현해보기/optional.ts
export class Optional<T> {
  private static throwEmptyError(): never {
    throw new Error('value is not defined');
  }
  public static of<S>(value: S): Optional<S> {
    if (value === undefined || value === null) Optional.throwEmptyError();
    return new Optional<S>(value);
  }
  public static ofNullable<S>(value?: S): Optional<S> {
    return new Optional<S>(value);
  }

  public static empty<S>(): Optional<S> {
    return new Optional<S>();
  }

  private constructor(private readonly value?: T) {}

  public get(): T {
    if (this.value === undefined || this.value === null) Optional.throwEmptyError();
    return this.value;
  }

  public orElse(other: T): T {
    if (this.value === undefined || this.value === null) return other;
    return this.value;
  }

  public orElseGet(fn: () => T): T {
    if (this.value === undefined || this.value === null) return fn();
    return this.value;
  }

  public map<S>(fn: (value: T) => S): Optional<S> {
    if (this.value === undefined || this.value === null) return Optional.empty<S>();
    return Optional.of(fn(this.value));
  }

  public filter(fn: (value: T) => boolean): Optional<T> {
    if (this.value === undefined || this.value === null || !fn(this.value)) return Optional.empty<T>();
    return Optional.empty<T>();
  }

  public isPresent() {
    return this.value !== undefined && this.value !== null;
  }

  public ifPresent(fn: () => void) {
    if (this.value !== undefined && this.value !== null) fn();
  }

  public ifPresentOrElse(fn: () => void, elseFn: () => void) {
    this.value !== undefined && this.value !== null ? fn() : elseFn();
  }

  public ifAbsent(fn: () => void) {
    if (this.value === undefined || this.value === null) fn();
  }

  public orElseThrow(fn: Error | (() => never)) {
    if (this.value === undefined || this.value === null) {
      if (fn instanceof Error) throw fn;
      fn();
    }
  }
}

File: 공부/구현해보기/optional.ex
defmodule Optional do
  defstruct value: nil, reason: nil

  def of({:ok, value}), do: %Optional{value: value}
  def of({:error, reason}), do: %Optional{value: nil, reason: reason}
  def of(%Optional{} = optional), do: optional
  def of(value), do: %Optional{value: value}

  def or_else(%Optional{value: nil}, func) when is_function(func, 0), do: func.() |> of()
  def or_else(%Optional{value: nil}, value), do: value |> of()
  def or_else(%Optional{} = optional, _), do: optional

  def is_present?(%Optional{value: nil}), do: false
  def is_present?(%Optional{}), do: true

  def is_nil?(%Optional{value: nil}), do: true
  def is_nil?(%Optional{}), do: false

  def if_present(%Optional{value: nil}, func) when is_function(func, 0), do: nil

  def if_present(%Optional{value: _}, func) when is_function(func, 0),
    do: (func.() && nil) || nil

  def if_present(%Optional{value: nil}, func) when is_function(func, 1), do: nil

  def if_present(%Optional{value: value}, func) when is_function(func, 1),
    do: (func.(value) && nil) || nil

  def filter(%Optional{value: nil} = optional, func) when is_function(func, 1), do: optional

  def filter(%Optional{value: value} = optional, func) when is_function(func, 1),
    do: if(func.(value), do: optional, else: %Optional{})

  def map(%Optional{value: nil} = optional, func) when is_function(func, 1), do: optional

  def map(%Optional{value: value}, func) when is_function(func, 1),
    do: func.(value) |> of()

  def get!(%Optional{value: nil, reason: reason}) when not is_nil(reason), do: raise(reason)
  def get!(%Optional{value: value}) when not is_nil(value), do: value

  def get(%Optional{value: value}), do: value

  def get(%Optional{value: nil}, func) when is_function(func, 0), do: func.()
  def get(%Optional{value: nil}, value), do: value
  def get(%Optional{value: value}, _), do: value

  def get_result(%Optional{value: nil, reason: nil}), do: {:error, :empty_optional}
  def get_result(%Optional{value: nil, reason: reason}), do: {:error, reason}
  def get_result(%Optional{value: value}), do: {:ok, value}

  def get_result(%Optional{value: nil}, func) when is_function(func, 0), do: {:ok, func.()}
  def get_result(%Optional{value: nil}, value), do: {:ok, value}
  def get_result(%Optional{value: value}, _), do: {:ok, value}
end
